SNOW-1256994: Getting started with native apps tutorial CLI update (#1174)
Snowflake-Labs/sfquickstarts: site/sfguides/src/getting_started_with_native_apps/getting_started_with_native_apps.md

### What You’ll Need 
- [VSCode](https://code.visualstudio.com/download) Installed

> - [Snowlflake CLI](https://docs.snowflake.com/developer-guide/snowflake-cli-v2/installation/installation) Installed

Reviewer: ```suggestion
- [Snowflake CLI](https://docs.snowflake.com/developer-guide/snowflake-cli-v2/installation/installation) Installed
```

To create our Snowflake Native Application, we will first clone the [starter project](https://github.com/Snowflake-Labs/sfguide-getting-started-with-native-apps) by running this command:

```bash

> snow app init --template git@github.com:Snowflake-Labs/sfguide-getting-started-with-native-apps.git

Reviewer: I suggest that we keep this as a `git clone` as we do not need the template expansion done by `snow app init`. As long as the repo itself has the `snowflake.yml` in it we're good

```

> This command is going to create an application project in your local machine, and it is going to do it from the template obatained from the github repo.

Reviewer: ```suggestion
This command is going to create an application project folder on your local machine, cloned from the github repo.
```

> If you do not have snow CLI or Git installed, you can also download the code directly from [GitHub](https://github.com/Snowflake-Labs/sfguide-getting-started-with-native-apps/archive/refs/heads/main.zip) and extract it to a local folder.

Reviewer: ```suggestion
If you do not have Snowflake CLI or Git installed, you can also download the code directly from [GitHub](https://github.com/Snowflake-Labs/sfguide-getting-started-with-native-apps/archive/refs/heads/main.zip) and extract it to a local folder.
```

        |   |   |-- udf.py
        |   |-- scripts
        |       |-- setup.sql

>                 |-- setup-package-script.sql

Reviewer: As in the other PR, let's make sure this file doesn't make it to the stage.

  order_id NUMBER(38,0), 
  ship_order_id NUMBER(38,0),
  status VARCHAR(60),
  lat FLOAT,
  lon FLOAT,
  duration NUMBER(38,0)
);"
```

### Upload Consumer Supply Chain Data

In this scenario, consumers will provider their own supply chain data (orders and site recovery data) from their own Snowflake account. The app will use the consumer's data to render graphs representing different aspects of the supply chain.

We'll use the `NATIVE_APP_QUICKSTART_DB` to store the consumer supply chain data.

```
snow sql -q "
USE WAREHOUSE NATIVE_APP_QUICKSTART_WH;

-- this database is used to store our data
USE DATABASE NATIVE_APP_QUICKSTART_DB;

USE SCHEMA NATIVE_APP_QUICKSTART_SCHEMA;

CREATE OR REPLACE TABLE MFG_ORDERS (
  order_id NUMBER(38,0), 
  material_name VARCHAR(60),
  supplier_name VARCHAR(60),
  quantity NUMBER(38,0),
  cost FLOAT,
  process_supply_day NUMBER(38,0)
);

CREATE OR REPLACE TABLE MFG_SITE_RECOVERY (
  event_id NUMBER(38,0), 
  recovery_weeks NUMBER(38,0),
  lat FLOAT,
  lon FLOAT
);"
```

Once we have created the necessary tables in the provider and consumer side, the next step is to load the csv files to the corresponding tables.

> That is accomplished using the `snow object stage copy` command, this uploads it directly to the table associated stage.

Reviewer: ```suggestion
That is accomplished using the `snow stage copy` command, which uploads it directly to the table associated stage.
```

```bash
# loading shipping data into table stage

> snow object stage copy ./app/data/shipping_data.csv @%MFG_SHIPPING --database NATIVE_APP_QUICKSTART_DB --schema NATIVE_APP_QUICKSTART_SCHEMA

Reviewer: Hot off the presses, this command is now deprecated. You can replace it with `snow stage copy` (just remove `object` for all stage commands). Make sure you have the latest version of Snowflake CLI installed

# loading orders data into table stage

> snow object stage copy ./app/data/order_data.csv @%MFG_ORDERS --database NATIVE_APP_QUICKSTART_DB --schema NATIVE_APP_QUICKSTART_SCHEMA

Reviewer: Same as above

# loading site recovery data into table stage

> snow object stage copy ./app/data/site_recovery_data.csv @%MFG_SITE_RECOVERY --database NATIVE_APP_QUICKSTART_DB --schema NATIVE_APP_QUICKSTART_SCHEMA

Reviewer: Same as above

```

Now the data is in the table stage, but its not yet loaded inside the table, for that, the following commands are run in the file:

```
snow sql -q"USE WAREHOUSE NATIVE_APP_QUICKSTART_WH;
-- this database is used to store our data
USE DATABASE NATIVE_APP_QUICKSTART_DB;

USE SCHEMA NATIVE_APP_QUICKSTART_SCHEMA;

COPY INTO MFG_SHIPPING
FILE_FORMAT = (TYPE = CSV
FIELD_OPTIONALLY_ENCLOSED_BY = '\"');

COPY INTO MFG_ORDERS
FILE_FORMAT = (TYPE = CSV
FIELD_OPTIONALLY_ENCLOSED_BY = '\"');

COPY INTO MFG_SITE_RECOVERY
FILE_FORMAT = (TYPE = CSV
FIELD_OPTIONALLY_ENCLOSED_BY = '\"');
"
```

## Share the Provider Shipping Data
Duration: 2

In order for this data to be available to the application consumer, we'll need to share it in the application package via reference usage.

> The following code is managed by the **setup-packge-script** file and its automatically run when we deploy the application. Here's what the SQL does:

Reviewer: ```suggestion
The following steps are performed by the **setup-package-script.sql** file, which is automatically run whenever we deploy the application:
```

- Creates a schema in the application package that will be used for sharing the shipping data

- Create a view within that schema

- Grants usage on the schema to the application package

- Grants reference usage on the database holding the provider shipping data to the application package

> - Grants SELECT privileges on the view to the application package, meaning the app will be able to SELECT on the view once it is installed

Reviewer: Does the spacing end up looking correct with a space between each bullet?

grant usage on procedure app_instance_schema.update_reference(string, string, string) to application role app_instance_role;
```

<!-- ------------------------ -->
## Create App Package
Duration: 2

A Snowflake Application Package is conceptually similar to that of an application installer for a desktop computer (like `.msi` for Windows or `.pkg` for Mac). An app package for Snowflake contains all the material used to install the application later, including the setup scripts. In fact, we will be using this app package in future steps to test our app!

Now that we've understood our project files, lets create the Snowflake Application Package so we can upload our project:

To create an application package you can go through the manual UI process or you can run:

```

> snowflake app run

Reviewer: ```suggestion
snow app run
```

Let's add a new function that simply outputs "Hello World!"

To do this, copy and paste the code below into **UDF.py**:

```python
def hello_world():
   return "Hello World!"
```

In the next step, we will expose this function to Consumers by adding it to our installation script.

<!-- ------------------------ -->
## Update Installation Script
Duration: 3

Let's add the following code snippet to our `setup.sql` script so we can use the python function we created in the previous step:

```python
create or replace function app_instance_schema.hello_world()
returns string
language python
runtime_version = '3.8'
packages = ('snowflake-snowpark-python')
imports = ('/libraries/udf.py')
handler = 'udf.hello_world';

grant usage on function app_instance_schema.hello_world() to application role app_instance_role;
```

<!-- ------------------------ -->
## Create App Package Version
Duration: 4

We've now added some basic functionality to the native app. From here, we will create the first version of our application package. You can have multiple versions available. 

> Yo create our application package version we are going to execute the following command:

Reviewer: ```suggestion
To create a version in our application package, execute the following command:
```

```

> snow app version create V1 --interactive --database NATIVE_APP_QUICKSTART_PACKAGE

Reviewer: Is the `--database` necessary here?

Author: You are right, I just tested it and it is not necessary

```

This code is going to create the Version number 1 of out application package, in the NATIVE_APP_QUICKSTART_PACKAGE 

> database, the --interactive directive is used to show more information in other prompts different than terminal devices.

Reviewer: I think `--interactive` might be the default for regular shells and could be skipped. I'm not sure the explanation helps the user's understanding as they're trying to learn something completely different in this moment.

Author: You are right, I just tested it and it is not necessary

> As we explained before, you can also set the version information inside the manifest.yml, therefore when you run the `snow app run` command, it is going to automatically create the versioned application. It would look something like this:

Reviewer: This isn't what happens; `snow app run` will only ever _use_ a version (not create one), and only if you pass in `--version` and `--patch` commands. Otherwise it uses stage dev mode (versionless).

```
version:
  name: V1
  label: Version One
  comment: The first version of the application
```

> In this case, because the first version was already created, the system is going to add a patch version by default to make the differentiation from both packages, but this could be explicitly set in the `snow app version create` options.

Reviewer: ```suggestion
Because the first version was already created, the system is going to add a new patch in the existing version. You can explicitly number the patch by adding `--patch <number>` to the `snow app version create` arguments.
```

To create a version in our application package, execute the following command:

```
snow app version create V1
```

> The code above is going to create the Version number 1 of out application package, in the NATIVE_APP_QUICKSTART_PACKAGE 

Reviewer: ```suggestion
Executing this command will create a version `V1` in our application package.
```

database.. You can explicitly number the patch by adding `--patch <number>` to the `snow app version create` arguments. Because the first version was already created, and because we did not specify the patch number, the system is going to add a new patch in the existing version.

> As we explained before, you can also set the version information inside the manifest.yml, therefore when you run the `snow app run` command, it is going to automatically use that version. It would look something like this:

Reviewer: ```suggestion
Executing this command will create a version `V1` in our application package.
You can explicitly number the patch by adding `--patch <number>` as an argument to `snow app version create`. Because this version does not yet exist and because we did not specify the patch number, the system is going to add our code as patch 0 in this version.

As explained before, you can also set the version information inside of manifest.yml. In that case, running `snow app version create` without any arguments will automatically use that version. It would look something like this:
```

-- This executes "setup.sql"; This is also what gets executed when installing the app
CREATE APPLICATION NATIVE_APP_QUICKSTART_APP FROM application package NATIVE_APP_QUICKSTART_PACKAGE using version V1 patch 0;
```

> When it finishes running, it is going to show you a link to open the app directly, or alternatively you should see the app NATIVE_APP_QUICKSTART_APP listed under Apps in the UI.

Reviewer: ```suggestion
When it finishes running, it is going to show you a link to open the app directly. Alternatively, you can see the NATIVE_APP_QUICKSTART_APP listed under Apps in the UI.
```

<!-- ------------------------ -->
## Run the Streamlit App
Duration: 1

Click on the app to launch it and give it a few seconds to warm up. 

When running the app for the first time, you'll be prompted to do some first-time setup by granting the app access to certain tables (i.e., create object-level bindings). The bindings link references defined in the manifest file to corresponding objects in the Snowflake account. These bindings ensure that the application can run as intended.

Upon running the application, you will see this:

![streamlit](assets/streamlit.png)

### Teardown

To delete the application you just created, you can run this command:

`snow app teardown`

> And to delete the database you used to poblate the application, this command will do:

Reviewer: ```suggestion
And to delete the database you used to populate the application, this command will do:
```
